File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.java
class Trie {
    Trie[] children;
    boolean isWord;
    public Trie() {
        children = new Trie[26];
        isWord = false;
    }
    
    public void insert(String word) {
        Trie node = this;
        for(int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);
            if(node.children[c - 'a'] == null)
                node.children[c-'a'] = new Trie();
            node = node.children[c - 'a'];
        }
        node.isWord = true;
    }
    
    public boolean search(String word) {
        Trie node = searchPrefix(word);
        return node != null && node.isWord;
    }
    
    public Trie searchPrefix(String word) {
        Trie node = this;
        for(char c : word.toCharArray()) {
            if(node.children[c - 'a'] == null) return null;
            node = node.children[c -'a'];
        }
        return node;
    }
    
    public boolean startsWith(String prefix) {
        return searchPrefix(prefix) != null;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 208-implement-trie-prefix-tree/208-implement-trie-prefix-tree.py
class Trie:

    def __init__(self):
        self.root = Node()

    def insert(self, word: str) -> None:
        curr = self.root
        for c in word:
           curr = curr.children[c]
        curr.is_word= True
        

    def search(self, word: str) -> bool:
        curr = self.root
        for c in word:
            if not curr.children.get(c):
                return False
            curr = curr.children[c]
        return curr.is_word

    def startsWith(self, prefix: str) -> bool:
        curr = self.root
        for c in prefix:
            if not curr.children.get(c):
                return False
            curr = curr.children[c]
        return True
    
class Node:
    def __init__(self):
        self.children = defaultdict(Node)
        self.is_word = False

# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)